### PR TITLE
issue #30 /tags should give a list of tags

### PR DESCRIPTION
**Pull request by voila**
A lot of edits are a result of running ocp-indent, and just change the indentation... I hope that's OK.

### CODE REVIEW

**Comment by hannesm** [comment]
cool. would be easier to review if you'd separate the ocp-indent in a separate commit (same PR is fine)

**Comment by voila** [comment]
Not sure if this fixed the indentation problem.

I have annotated the relevant changes, in the commit, to make it easier to review.

I am unsure of what I should do re:code formating, in the future... I use Emacs `M-x untabify` because some files in the master branch have tabs in them, but that also seem to change the indentation. So I used ocp-indent as @Engil suggested, but the resulting indentation is different too, adding a lot of noise to my commits.

File: canopy_content.ml [hunk 1]
```diff
@@ -49,14 +49,24 @@ let to_atom = function
...
-     m.Canopy_article.created
+    m.Canopy_article.created
 
 let compare a b = Ptime.compare (date b) (date a)
 
 let updated = function
   | Markdown m ->
     m.Canopy_article.updated
+
+let tags content_map =
```
Comment (voila) [inline canopy_content.ml:64]: a new function that returns all existings tags (as string list) from the cache map (populated in Store.fill_cache)

File: canopy_article.ml [hunk 2]
```diff
@@ -49,14 +49,28 @@ let to_tyxml_listing_entry article =
...
-    ] in
+    h4 ~a:[a_class ["list-group-item-heading"]] [pcdata article.title];
+    span ~a:[a_class ["author"]] [pcdata author];
+    pcdata " ";
+    pcdata "("; time [pcdata created]; pcdata ")";
+    br ();
+  ] in
   a ~a:[a_href article.uri; a_class ["list-group-item"]] (content ++ abstract)
 
+let to_tyxml_tags tags =
```
Comment (voila) [inline canopy_article.ml:60]: A function to format the list of tags as links to related articles, as HTML. This is used to generates the page for the `/tags` URI

File: canopy_dispatch.ml [hunk 3]
```diff
@@ -66,67 +66,72 @@ module Make (S: Cohttp_lwt.Server) (C: V1_LWT.CONSOLE) (Disk: V1_LWT.KV_RO)
     | uri::[] when uri = config.Canopy_config.push_hook_path ->
       store.update () >>= fun l ->
       respond_update l
+    | "tags"::[] -> (
```
Comment (voila) [inline canopy_dispatch.ml:69]: The dispatch rules for the `/tags` URI

File: canopy_dispatch.ml [hunk 4]
```diff
@@ -66,67 +66,72 @@ module Make (S: Cohttp_lwt.Server) (C: V1_LWT.CONSOLE) (Disk: V1_LWT.KV_RO)
...
+      let tags = Canopy_content.tags !cache in
+      let content = Canopy_article.to_tyxml_tags tags in
+      respond_html ~headers ~title:config.Canopy_config.blog_name ~content ~updated
```
Comment (hannesm) [inline canopy_dispatch.ml:73]: hmm, the `updated` here is the Canopy startup time... but tags might be added later (by addng a new article, or new tags to an existing article).  should we either drop or have the `post` hook not only refill the `!cache`, but also remember the timestamp (to be used only here (and maybe for the atom feed))?

Comment (voila) [inline canopy_dispatch.ml:73]: good spotting... I guess it is silly to recalculate this from the cache, every time...
Should the cache be extended to also store the last tag timestamp, the last article timestamp, the list of tags, etc... ?

Comment (hannesm) [inline canopy_dispatch.ml:73]: this is not very clear to me... atm the cache contains the articles, but no caching of listings or tags is done... maybe they should be cached (as you might have noticed [here](https://hannes.nqsb.io/Posts/nqsbWebsite) I prefer to have as little done as possible on each individual request)?

For this change request here, I'd just use the `store.last_commit () >>= fun updated`, the same used for the atom feed.  Caching can come in a separate PR, but @Engil should tell his opinion :)

Comment (voila) [inline canopy_dispatch.ml:73]: I am happy to change the code to use `store.last_commit () >>= fun updated` for this PR, but what if the last commit was not a new tag but a new article ? Shouldn't the Etag (for the `/tags` page) reflects the set of existing tags only ?

Comment (hannesm) [inline canopy_dispatch.ml:73]: yep, that's an conservative approximation and will result in more data being transferred in the worst case than needed... but better than an underapproximation which leads to outdated data on the clients.. ;)  a new article might as well introduce new tags, as might a change to an existing article.. the logic how to get the last update to any tag (add/remove) is non-trivial

File: canopy_dispatch.ml (hunk 3 above)
Comment (voila) [inline canopy_dispatch.ml]: Now with `updated` coming from `store.last_commit ()`

File: canopy_dispatch.ml (hunk 4 above)
Comment (voila) [inline canopy_dispatch.ml]: Sounds good. I have used `store.last_commit ()` for `updated`

Comment (voila) [inline canopy_dispatch.ml:73]: Sounds good to me :)

**Comment by hannesm** [comment]
lgtm

File: canopy_dispatch.ml (hunk 4 above)
Comment (abbysmal) [inline canopy_dispatch.ml:73]: I think at some point we should improve the cache mechanism (which is a Map other Canopy_content.t right now), abstracting thing in such a way we can cache easily other things than articles.
I'm not too sure on how to do that properly, for now this solution works fine enough, we'll see later for something more clever

**Comment by abbysmal** [comment]
Everything looks fine to me, tested it just now, looks great! (was a bit busy this week end again, sorry for being late to the party…)
Thank you @hannesm for reviewing everything, and thank you @voila again for delivering such nice PRs…
Merging. :)

**Pull request merged**